File: Enclave-macOS/Sources/EnclaveIO.swift
import Foundation

enum EnclaveIO {
    static func expandedPath(_ path: String) -> String {
        (path as NSString).expandingTildeInPath
    }

    static func validatePlaintextInput(_ url: URL) throws {
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) else {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }
        guard FileManager.default.isReadableFile(atPath: url.path) else {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }
        guard !EnclaveCrypto.isArchiveFilename(url.lastPathComponent) else {
            throw EnclaveError.cannotEncryptArchive
        }
        if isDirectory.boolValue {
            guard directoryContainsRegularFile(url) else {
                throw EnclaveError.emptyFolder
            }
        } else {
            try validateFileSize(url)
        }
    }

    static func prepareEncryptionPayload(from url: URL) throws -> (data: Data, filename: String) {
        try validatePlaintextInput(url)
        if EnclaveFolder.isDirectory(url) {
            let data = try EnclaveFolder.pack(directory: url)
            return (data, EnclaveFolder.storedFilename(forDirectory: url))
        }
        let data = try Data(contentsOf: url)
        guard data.count <= EnclaveCrypto.maxPlaintextSize else {
            throw EnclaveError.payloadTooLarge
        }
        return (data, url.lastPathComponent)
    }

    static func resolveFolderDecryptParent(output: String?, relativeTo inputURL: URL) throws -> URL {
        guard let output else {
            return inputURL.deletingLastPathComponent()
        }

        var isDirectory: ObjCBool = false
        let exists = FileManager.default.fileExists(atPath: output, isDirectory: &isDirectory)
        if exists && isDirectory.boolValue {
            return URL(fileURLWithPath: output, isDirectory: true)
        }

        let parent = URL(fileURLWithPath: output).deletingLastPathComponent()
        if !parent.path.isEmpty, parent.path != "/" {
            try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
        }
        return parent.path.isEmpty ? inputURL.deletingLastPathComponent() : parent
    }

    static func writeDecryptedPayload(_ payload: Data, filename: String, to outputURL: URL) throws {
        if EnclaveFolder.isFolderPayload(payload) {
            let folderName = EnclaveFolder.displayFolderName(from: filename)
            try EnclaveFolder.unpack(payload: payload, to: outputURL, folderName: folderName)
            return
        }
        try payload.write(to: outputURL, options: .atomic)
    }

    static func validateArchiveInput(_ url: URL) throws {
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory),
              !isDirectory.boolValue else {
            throw EnclaveError.invalidFormat
        }
        guard FileManager.default.isReadableFile(atPath: url.path) else {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }

        try validateArchiveSize(url)

        let headerSize = EnclaveCrypto.magic.count
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }

        guard let header = try handle.read(upToCount: headerSize), EnclaveCrypto.isArchiveData(header) else {
            throw EnclaveError.invalidFormat
        }
    }

    static func readArchive(from url: URL) throws -> Data {
        try validateArchiveInput(url)
        let data = try Data(contentsOf: url)
        guard data.count <= EnclaveCrypto.maxArchiveSize else {
            throw EnclaveError.payloadTooLarge
        }
        return data
    }

    static func resolveOutputURL(
        output: String?,
        defaultName: String,
        relativeTo inputURL: URL,
        appendedFilename: String? = nil
    ) throws -> URL {
        guard let output else {
            return inputURL.deletingLastPathComponent().appendingPathComponent(defaultName)
        }

        var isDirectory: ObjCBool = false
        let exists = FileManager.default.fileExists(atPath: output, isDirectory: &isDirectory)
        if exists && isDirectory.boolValue {
            let name = appendedFilename ?? defaultName
            return URL(fileURLWithPath: output, isDirectory: true).appendingPathComponent(name)
        }

        let outputURL = URL(fileURLWithPath: output)
        let parent = outputURL.deletingLastPathComponent()
        if !parent.path.isEmpty, parent.path != "/" {
            try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
        }
        return outputURL
    }

    static func normalizedArchiveURL(_ url: URL) -> URL {
        guard EnclaveCrypto.isArchiveFilename(url.lastPathComponent) else {
            return url.deletingPathExtension().appendingPathExtension(EnclaveCrypto.fileExtension)
        }
        return url
    }

    private static func validateArchiveSize(_ url: URL) throws {
        let values = try url.resourceValues(forKeys: [.fileSizeKey, .isRegularFileKey, .isSymbolicLinkKey])
        if values.isSymbolicLink == true {
            throw EnclaveError.invalidFormat
        }
        if values.isRegularFile == false {
            throw EnclaveError.invalidFormat
        }

        var fileSize = values.fileSize
        if fileSize == nil {
            let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
            fileSize = attributes[.size] as? Int
        }
        guard let fileSize else {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }
        if fileSize > EnclaveCrypto.maxArchiveSize {
            throw EnclaveError.payloadTooLarge
        }
    }

    private static func validateFileSize(_ url: URL) throws {
        let values = try url.resourceValues(forKeys: [.fileSizeKey, .isRegularFileKey, .isSymbolicLinkKey])
        if values.isSymbolicLink == true {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }
        if values.isRegularFile == false {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }

        var fileSize = values.fileSize
        if fileSize == nil {
            let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
            fileSize = attributes[.size] as? Int
        }
        guard let fileSize else {
            throw EnclaveError.unreadableFile(url.lastPathComponent)
        }
        if fileSize > EnclaveCrypto.maxPlaintextSize {
            throw EnclaveError.payloadTooLarge
        }
    }

    private static func directoryContainsRegularFile(_ url: URL) -> Bool {
        guard let enumerator = FileManager.default.enumerator(
            at: url,
            includingPropertiesForKeys: [.isRegularFileKey, .isSymbolicLinkKey],
            options: [.skipsPackageDescendants, .skipsHiddenFiles]
        ) else {
            return false
        }

        for case let item as URL in enumerator {
            guard let values = try? item.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey]) else {
                continue
            }
            if values.isSymbolicLink == true {
                continue
            }
            if values.isRegularFile == true {
                return true
            }
        }
        return false
    }
}

File: Enclave For Windows/src/EnclaveCore/EnclaveIo.cs
using System.Text;

namespace EnclaveCore;

/// <summary>File IO, validation, and output-path resolution. Adapted from the
/// macOS URL-based version to Windows paths.</summary>
public static class EnclaveIo
{
    public static string ExpandedPath(string path)
    {
        if (path == "~")
            return Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (path.StartsWith("~/", StringComparison.Ordinal) || path.StartsWith("~\\", StringComparison.Ordinal))
            return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), path[2..]);
        return path;
    }

    public static void ValidatePlaintextInput(string path)
    {
        bool isDir = Directory.Exists(path);
        if (!isDir && !File.Exists(path))
            throw EnclaveException.UnreadableFile(Name(path));
        if (!IsReadable(path, isDir))
            throw EnclaveException.UnreadableFile(Name(path));
        if (EnclaveCrypto.IsArchiveFilename(Name(path)))
            throw EnclaveException.CannotEncryptArchive();

        if (isDir)
        {
            if (!DirectoryContainsRegularFile(path))
                throw EnclaveException.EmptyFolder();
        }
        else
        {
            ValidateFileSize(path);
        }
    }

    public static (byte[] Data, string Filename) PrepareEncryptionPayload(string path)
    {
        ValidatePlaintextInput(path);
        if (EnclaveFolder.IsDirectory(path))
        {
            byte[] packed = EnclaveFolder.Pack(path);
            return (packed, EnclaveFolder.StoredFilename(path));
        }

        byte[] data = File.ReadAllBytes(path);
        if (data.LongLength > EnclaveCrypto.MaxPlaintextSize)
            throw EnclaveException.PayloadTooLarge();
        return (data, Name(path));
    }

    public static string ResolveFolderDecryptParent(string? output, string inputPath)
    {
        if (output is null)
            return Path.GetDirectoryName(Path.GetFullPath(inputPath)) ?? Directory.GetCurrentDirectory();

        if (Directory.Exists(output))
            return Path.GetFullPath(output);

        string parent = Path.GetDirectoryName(Path.GetFullPath(output)) ?? "";
        if (!string.IsNullOrEmpty(parent))
            Directory.CreateDirectory(parent);
        return string.IsNullOrEmpty(parent)
            ? Path.GetDirectoryName(Path.GetFullPath(inputPath)) ?? Directory.GetCurrentDirectory()
            : parent;
    }

    public static void WriteDecryptedPayload(byte[] payload, string filename, string outputPath)
    {
        if (EnclaveFolder.IsFolderPayload(payload))
        {
            string folderName = EnclaveFolder.DisplayFolderName(filename);
            EnclaveFolder.Unpack(payload, outputPath, folderName);
            return;
        }
        File.WriteAllBytes(outputPath, payload);
    }

    public static void ValidateArchiveInput(string path)
    {
        if (!File.Exists(path))
            throw EnclaveException.InvalidFormat();
        if ((File.GetAttributes(path) & FileAttributes.ReparsePoint) != 0)
            throw EnclaveException.InvalidFormat();
        if (!IsReadable(path, isDirectory: false))
            throw EnclaveException.UnreadableFile(Name(path));

        long size = new FileInfo(path).Length;
        if (size > EnclaveCrypto.MaxArchiveSize)
            throw EnclaveException.PayloadTooLarge();

        Span<byte> header = stackalloc byte[EnclaveCrypto.Magic.Length];
        using var stream = File.OpenRead(path);
        int read = stream.Read(header);
        if (read < header.Length || !EnclaveCrypto.IsArchiveData(header))
            throw EnclaveException.InvalidFormat();
    }

    public static byte[] ReadArchive(string path)
    {
        ValidateArchiveInput(path);
        byte[] data = File.ReadAllBytes(path);
        if (data.LongLength > EnclaveCrypto.MaxArchiveSize)
            throw EnclaveException.PayloadTooLarge();
        return data;
    }

    public static string ResolveOutputPath(
        string? output,
        string defaultName,
        string inputPath,
        string? appendedFilename = null)
    {
        if (output is null)
        {
            string dir = Path.GetDirectoryName(Path.GetFullPath(inputPath)) ?? Directory.GetCurrentDirectory();
            return Path.Combine(dir, defaultName);
        }

        if (Directory.Exists(output))
            return Path.Combine(Path.GetFullPath(output), appendedFilename ?? defaultName);

        string parent = Path.GetDirectoryName(Path.GetFullPath(output)) ?? "";
        if (!string.IsNullOrEmpty(parent))
            Directory.CreateDirectory(parent);
        return Path.GetFullPath(output);
    }

    public static string NormalizedArchivePath(string path)
    {
        if (!EnclaveCrypto.IsArchiveFilename(Name(path)))
            return Path.ChangeExtension(path, "." + EnclaveCrypto.FileExtension);
        return path;
    }

    // ===================== helpers =====================

    private static string Name(string path)
    {
        string trimmed = path.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        string name = Path.GetFileName(trimmed);
        return string.IsNullOrEmpty(name) ? path : name;
    }

    private static bool IsReadable(string path, bool isDirectory)
    {
        try
        {
            if (isDirectory)
            {
                _ = Directory.EnumerateFileSystemEntries(path).Any();
                return true;
            }
            using var s = File.OpenRead(path);
            return true;
        }
        catch
        {
            return false;
        }
    }

    private static void ValidateFileSize(string path)
    {
        if ((File.GetAttributes(path) & FileAttributes.ReparsePoint) != 0)
            throw EnclaveException.UnreadableFile(Name(path));
        long size = new FileInfo(path).Length;
        if (size > EnclaveCrypto.MaxPlaintextSize)
            throw EnclaveException.PayloadTooLarge();
    }

    private static bool DirectoryContainsRegularFile(string path)
    {
        var stack = new Stack<string>();
        stack.Push(path);
        while (stack.Count > 0)
        {
            string dir = stack.Pop();
            string[] entries;
            try { entries = Directory.GetFileSystemEntries(dir); }
            catch { continue; }
            foreach (string entry in entries)
            {
                FileAttributes attr;
                try { attr = File.GetAttributes(entry); }
                catch { continue; }
                string name = Path.GetFileName(entry);
                if (name.StartsWith('.')) continue;
                if ((attr & FileAttributes.Hidden) != 0) continue;
                if ((attr & FileAttributes.ReparsePoint) != 0) continue;
                if ((attr & FileAttributes.Directory) != 0) { stack.Push(entry); continue; }
                return true;
            }
        }
        return false;
    }
}
